File: Activity-4/src/core/Packet.java
package core;

import java.nio.ByteBuffer;

/**
 * Classe representante da estrutura de dados utilizada para
 * fazer as operações com o servidor.
 * 
 * @author daniel e mara
 */
public class Packet {
    private int mode;
    private int verb;
    private int finished;
    private int formatData;
    private String resource = "";
    private String params = "";
    private byte[] data = new byte[0];

    public Packet mode(int m) {
        mode = m;
        return this;
    }
    public Packet verb(int v) {
        verb = v;
        return this;
    }
    public Packet finished(int f) {
        finished = f;
        return this;
    }
    public Packet formatData(int f) {
        formatData = f;
        return this;
    }
    public Packet resource(String r) {
        resource = r;
        return this;
    }
    public Packet params(String p) {
        params = p;
        return this;
    }
    public Packet data(byte[] d) {
        data = d;
        return this;
    }
    private Packet resource(ByteBuffer b) {
        byte[] r = new byte[b.get()];
        b.get(r);
        resource = new String(r);
        return this;
    }
    private Packet params(ByteBuffer b) {
        byte[] p = new byte[b.get()];
        b.get(p);
        params = new String(p);
        return this;
    }
    private Packet data(ByteBuffer b) {
        byte[] d = new byte[b.getShort()];
        b.get(d);
        data = d;
        return this;
    }

    public int mode() {
        return mode;
    }
    public int verb() {
        return verb;
    }
    public int finished() {
        return finished;
    }
    public int formatData() {
        return formatData;
    }
    public String resource() {
        return resource;
    }
    public String params() {
        return params;
    }
    public byte[] data() {
        return data;
    }
    
    public int length() {
        return 8 + this.resource.length() + this.params.length() + this.data.length;
    }

    public byte[] getBytes() {
        return ByteBuffer.allocate(this.length())
            .put((byte) mode)
            .put((byte) verb)
            .put((byte) finished)
            .put((byte) formatData)
            .put((byte) resource.length())
            .put(resource.getBytes())
            .put((byte) params.length())
            .put(params.getBytes())
            .putShort((byte) data.length)
            .put(data)
            .array();
     }

    public static Packet fromBytes(byte[] buffer) {
        ByteBuffer b = ByteBuffer.wrap(buffer).asReadOnlyBuffer();
        return new Packet()
            .mode(b.get())
            .verb(b.get())
            .finished(b.get())
            .formatData(b.get())
            .resource(b)
            .params(b)
            .data(b);
    }

    public String toString() {
        return mode + "|" + verb + "|" + resource + "|" + params + "|" 
            + finished + "|" + formatData + "|" + data.length;
    }
}

File: Activity-4/Python/core/packet.py
import bitstring


class Packet(object):
    '''
        Classe representante da estrutura de dados utilizada para
    fazer as operações com o cliente.

    Argumentos e atributos:
        mode        (int): Request (1), Response (2)
        verb        (int): GET (1), POST (2), PUT (3), DELETE (4), ERROR (5)
        finished    (int): False (0), True (1)
        format_data (int): AUTO (0), JSON (1), PROTOCOL BUFFER (2)
        resource    (str) [max len 255]
        params      (str) [max len 255]
        data        (bytearray)
    '''
    def __init__(
        self,
        mode: int,
        verb: int,
        finished: int = 1,
        format_data: int = 1,
        resource: str = "",
        params: str = "",
        data: bytearray = bytearray()
    ):
        self.mode = mode
        self.verb = verb
        self.finished = finished
        self.format_data = format_data
        self.resource = resource
        self.params = params
        self.data = data

    def __str__(self):
        end = f"|{self.resource}|{self.params}|{self.finished}|{self.format_data}|{len(self.data)}"
        try:
            mode = ["REQ", "RES"][self.mode-1]
            verb = ["GET", "POST", "PUT", "DELETE"][self.verb-1]
            return f"{mode}|{verb}" + end
        except:
            return f"{self.mode}|{self.verb}" + end 

    def length(self):
        length_data = len(self.data) if self.data != None else 0 
        return 7 + len(self.resource) + len(self.params) + length_data

    def getBytes(self):
        stream = bitstring.BitStream()
        stream.append(f"int:8={self.mode}")
        stream.append(f"int:8={self.verb}")
        stream.append(f"int:8={self.finished}")
        stream.append(f"int:8={self.format_data}")
        stream.append(f"int:8={len(self.resource)}")
        stream.append(bytearray(map(ord, self.resource)))
        stream.append(f"int:8={len(self.params)}")
        stream.append(bytearray(map(ord, self.params)))
        stream.append(f"int:16={len(self.data)}")
        stream.append(self.data)
        return stream.bytes
    
    def fromBytes(data: bytearray):
        stream = bitstring.BitStream(data)
        mode = stream[:8].uint; del stream[:8]
        verb = stream[:8].uint; del stream[:8]
        finished = stream[:8].uint; del stream[:8]
        format_data = stream[:8].uint; del stream[:8]

        length = stream[:8].uint; del stream[:8]
        print("DEBUG resource: ", length)
        resource = stream[:8*length].bytes.decode("UTF8")
        del stream[:8*length]

        length = stream[:8].uint; del stream[:8]
        print("DEBUG params: ", length)
        params = stream[:8*length].bytes.decode("UTF8")
        del stream[:8*length]

        length = stream[:16].uint; del stream[:16]
        print("DEBUG data: ", length)
        return Packet(mode, verb, finished, format_data, resource, params, stream.bytes)


if __name__ == "__main__":
    p = Packet(1, 2, "alunos")
    p.params = "periodo=1&cod_curso=1"
    p.data = bytearray(map(ord, "Ola mundo!"))

    data = p.getBytes()
    print(p)
    print(data)

    k = Packet.fromBytes(data)
    data = k.getBytes()
    print(k)
    print(data)
    